File: C++/128. LongestConsecutiveSequence.cpp
/*
 Given an unsorted array of integers, find the length of the longest consecutive elements sequence.

Your algorithm should run in O(n) complexity.

Example:

Input: [100, 4, 200, 1, 3, 2]
Output: 4
Explanation: The longest consecutive elements sequence is [1, 2, 3, 4]. Therefore its length is 4.
 */

class Solution {
public:
    int longestConsecutive(vector<int> &nums) {

        // initialising a has_set to store the values that we have already seen
        unordered_set<int> hash_set;

        // the final resul that we have to return
        int largest_sequence = 0;

        // going through vector and storing the values in the hash set
        for (int i = 0; i < nums.size(); i++)
            hash_set.insert(nums[i]);

        // going through the vector again and computing the sequence
        for (int i = 0; i < nums.size(); i++) {
            int current_num = nums[i];
            int short_sequence = 1;

            // finding if the current_num +1 and curent_num -1 are present in the hash set or not
            if (hash_set.find(current_num - 1) == hash_set.end()) {
                while (hash_set.find(current_num + 1) != hash_set.end()) {
                    current_num++;
                    short_sequence++;
                }
            }

            //finding the maximum between the local sequence and the global sequence
            largest_sequence = max(largest_sequence, short_sequence);

        }


        //returning the largest_sequences
        return largest_sequence;

    }
};

File: Java/128. Longest Consecutive Sequence.java
Problem: Given an unsorted array of integers nums, return the length of the longest consecutive elements sequence.

Example 1:

Input: nums = [100,4,200,1,3,2]
Output: 4
Explanation: The longest consecutive elements sequence is [1, 2, 3, 4]. Therefore its length is 4.
Example 2:

Input: nums = [0,3,7,2,5,8,4,6,0,1]
Output: 9


The idea here is to store all the numbers in a set for easy access. For each number checking whether the subsequent number is present.
Increasing the length depending on the numbers present.

Code:

class Solution {
    public int longestConsecutive(int[] nums) {
        //To store the list of all numbers
        HashSet<Integer> set = new HashSet<>();
        for(int a:nums)
            set.add(a);
	// used to get the longest length of consecutive length
        int ans = 0;
        for(int i =0;i<nums.length;i++){
	    // To check if the current number is already counted	
            if(!set.contains(nums[i]-1)){
                int j = nums[i];
		//Incrementing the count till the subsequent numbers are found
                while(set.contains(j))
                    j++;
		//Storing the longest length.
                ans = Math.max(ans, j-nums[i]);
            }
        }
        return ans;
    }
}